frontend/ui/widget/qrwidget.lua

```diff
@@ -4,34 +4,43 @@ QRWidget shows a QR code for a given text.
 
 local ImageWidget = require("ui/widget/imagewidget")
 local RenderImage = require("ui/renderimage")
-local Widget = require("ui/widget/widget")
 local logger = require("logger")
 local qrencode = require("qrencode")
 
-local QRWidget = Widget:new{
+local QRWidget = ImageWidget:extend{
     scale_factor = 0,
     text = ""
     -- see ImageWidget for other options.
 }
 
 function QRWidget:init()
-    local ok, ret = qrencode.qrcode(self.text)
+    local ok, ret = qrencode.qrcode(self.text:sub(1, 2953))
     if not ok then
         logger.info("QRWidget: failed to generate QR code.")
         return
     else
         -- We generate pbm data
-        local qr = {"P1\n", #ret[1], " ", #ret, "\n"}
+        local qr = {"P1\n", 2*#ret[1], " ", 2*#ret, "\n"}
         for _, col in ipairs(ret) do
-            for _, lgn in ipairs(col) do
-                table.insert(qr, lgn > 0 and "1 " or "0 ")
+            for i = 1, 2 do
```

Comment:
(I'd also make the scaling factor a variable to make the intent slightly clearer, here and in the MUL above).